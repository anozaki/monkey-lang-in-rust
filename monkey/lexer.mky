let five = 5;
let ten = 10;

let add = fn(x, y) {
  x + y;
};

let result = add(five, ten);
!-/*5;
5 < 10 > 5;

if (5 < 10) {
    return true;
} else {
    return false;
}

5 == 4;
5 != 4;
5 <= 6;
5 >= 7;

fn() {

}

"this is a test";
let a = "this is another\n\"test\"";
[1, 2, 3];
[1, 2 * 3, 5];

{foo: 5, a: "test"}
