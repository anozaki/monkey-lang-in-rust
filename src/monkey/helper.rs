pub fn is_identifier(ch: char) -> bool {
    ch.is_ascii_alphabetic() || ch == '_'
}

pub fn is_digit(ch: char) -> bool {
    ch.is_ascii_digit()
}

pub fn is_whitespace(ch: char) -> bool {
    ch.is_whitespace()
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_is_identifier() {
        assert_eq!(is_identifier('a'), true);
        assert_eq!(is_identifier('_'), true);
        assert_eq!(is_identifier('あ'), false);
        assert_eq!(is_identifier('1'), false);
    }

    #[test]
    fn test_is_digit() {
        assert_eq!(is_digit('1'), true);
        assert_eq!(is_digit('2'), true);
        assert_eq!(is_digit('3'), true);
        assert_eq!(is_digit('4'), true);
        assert_eq!(is_digit('_'), false);
        assert_eq!(is_digit('%'), false);
        assert_eq!(is_digit('１'), false); // non ascii 1
    }

    #[test]
    fn test_is_whitespace() {
        assert_eq!(is_whitespace(' '), true);
        assert_eq!(is_whitespace('\n'), true);
        assert_eq!(is_whitespace('\t'), true);
        assert_eq!(is_whitespace('　'), true); // non-ascii space
    }
}