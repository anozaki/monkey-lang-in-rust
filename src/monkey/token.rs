use std::fmt::{Debug, Formatter};

#[derive(Debug, Copy, Clone, PartialEq)]
pub enum TokenType {
    Illegal,
    EndOfFile,

    Comment,

    String,

    // Identifier/Literals
    Identifier,
    Integer,

    // Operators
    Assign,
    Plus,
    Minus,
    Bang,
    Asterisk,
    Slash,

    LessThan,
    GreaterThan,

    Equal,
    NotEqual,
    LessThanEqual,
    GreaterThanEqual,

    // Delimiters
    Comma,
    Colon,
    Semicolon,

    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,

    // Keywords
    Function,
    Let,
    True,
    False,
    If,
    Else,
    Return,
}

#[derive(PartialEq, Clone)]
pub struct Span {
    pub start: usize,
    pub end: usize,
    pub line_start: usize,
    pub line_end: usize,
}

impl Debug for Span {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f, "[L{}-{:?}:{:?}]",
            self.line_start, self.start, self.end
        )
    }
}

#[derive(PartialEq, Clone)]
pub struct Token {
    pub token_type: TokenType,
    pub literal: String,

    pub span: Span,
}

impl Debug for Token {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f, "Token({:?}, {:?} {:?})",
            self.token_type, self.literal, self.span
        )
    }
}
