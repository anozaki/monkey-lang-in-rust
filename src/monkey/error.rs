use std::rc::Rc;
use thiserror::Error;
use crate::monkey::token::Token;

#[derive(Error, Debug, PartialEq, Clone)]
pub enum Error {
    #[error("Generic {0}")]
    TodoError(String),

    #[error("Invalid token found at {0:?}")]
    InvalidTokenError(Rc<Token>),

    #[error("Invalid escape char {0:?}")]
    InvalidCharError(char),
}