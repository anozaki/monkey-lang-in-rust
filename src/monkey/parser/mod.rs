use std::rc::Rc;

use crate::monkey::lexer::Lexer;
use crate::monkey::parser::ast::{Program, StatementNode};
use crate::monkey::Result;
use crate::monkey::token::{Token, TokenType};
use crate::monkey::token::TokenType::Semicolon;

pub mod parser;
pub mod expression;
pub mod ast;


#[macro_export]
macro_rules! try_next {
    ($self: ident, $token: expr) => {
        if $self.current().token_type != $token {
            return Err(Error::InvalidTokenError(Rc::clone($self.current())));
        }
        $self.next();
    }
}

fn next_token(lexer: &dyn Lexer) -> Token {
    match lexer.token() {
        Ok(token) => token,
        Err(message) => panic!("Unable to parse: {message}")
    }
}

pub struct Parser<'a> {
    lexer: &'a dyn Lexer,

    token_current: Rc<Token>,
    token_peek: Rc<Token>,
}

impl <'a> Parser <'a> {
    pub fn new(lexer: &'a dyn Lexer) -> Self {
        let token_current = Rc::new(next_token(lexer));
        let token_peek = Rc::new(next_token(lexer));
        Parser {
            lexer,
            token_current,
            token_peek,
        }
    }

    pub fn current(&self) -> &Rc<Token> {
        return &self.token_current;
    }

    pub fn peek(&self) -> &Rc<Token> {
        return &self.token_peek;
    }

    pub fn next(&mut self) {
        self.token_current = Rc::clone(&self.token_peek);
        self.token_peek = Rc::new(next_token(self.lexer));
    }

    pub fn skip_semicolon(&mut self) {
        while self.current().token_type == Semicolon {
            self.next()
        }
    }

    pub fn parse_program(&mut self) -> Result<Box<Program>> {
        let mut program: Box<Program> = Box::new(Program::default());

        while self.current().token_type != TokenType::EndOfFile && self.current().token_type != TokenType::RightBrace {
            program.statements.push(self.parse_statement()?);
            self.skip_semicolon();
        }

        Ok(program)
    }

    pub fn parse_statement(&mut self) -> Result<StatementNode> {
        Ok(match self.current().token_type {
            TokenType::Let => self.parse_let()?,
            TokenType::Return => self.parse_return()?,
            TokenType::If => self.parse_if()?,
            _ => self.parse_expression()?,
        })
    }
}

#[cfg(test)]
mod test {
    use std::fs;
    use crate::monkey::lexer::MonkeyLexer;

    use super::*;

    pub fn evaluate(source: &str) -> String {
        let contents = fs::read_to_string(source).unwrap();
        let lexer = MonkeyLexer::new(&contents);

        let mut parser = Parser::new(&lexer);
        let program = parser.parse_program().unwrap();

        let mut output = String::new();
        for statement in &program.statements {
            output += &format!("{:?}\n", statement);
        }

        output
    }

    macro_rules! evaluate {
        ($name:ident, $file:literal) => {
            #[test]
            fn $name() {
                let output = evaluate(&format!("monkey/{}", $file));
                insta::assert_snapshot!(output)
            }
        }
    }

    evaluate!(test_parse_let, "test_parser_let.mky");
    evaluate!(test_parse_return, "test_parser_return.mky");
    evaluate!(test_parser_infix, "test_parser_infix.mky");
    evaluate!(test_parser_expression, "test_parser_expression.mky");
    evaluate!(test_parser_expression_prec, "test_parser_expression_prec.mky");
    evaluate!(test_parser_if, "test_parser_if.mky");
    evaluate!(test_parser_function_literals, "test_parser_function_literals.mky");
    evaluate!(test_parser_closure, "test_parser_closure.mky");
    evaluate!(test_parser_string, "test_parser_string.mky");
    evaluate!(test_parser_index, "test_parser_index.mky");
    evaluate!(test_parser_hash, "test_parser_hash.mky");
}
