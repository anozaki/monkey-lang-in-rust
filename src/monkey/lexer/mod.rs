use std::cell::Cell;
use std::collections::VecDeque;
use std::fs;
use std::rc::Rc;
use crate::monkey::error::Error::{InvalidTokenError, TodoError, InvalidCharError};
use crate::monkey::Result;

use crate::monkey::helper::{is_digit, is_identifier, is_whitespace};
use crate::monkey::token::{Span, Token, TokenType};

pub trait Lexer {
    fn token(&self) -> Result<Token>;
}

pub struct MonkeyLexer {
    /// Input
    input: Vec<char>,

    column: Cell<usize>,
    position: Cell<usize>,
    read_position: Cell<usize>,
    line: Cell<usize>,
    ch: Cell<char>,
}

impl Lexer for MonkeyLexer {
    fn token(&self) -> Result<Token> {
        self.read()
    }
}


impl MonkeyLexer {
    pub fn new(input: &str) -> Self {
        let input: Vec<char> = input.chars().collect();
        let ch: char = input[0];
        Self {
            input,
            column: Cell::new(0),
            position: Cell::new(0),
            read_position: Cell::new(1),
            line: Cell::new(0),
            ch: Cell::new(ch.into()),
        }
    }

    fn peek(&self) -> char {
        self.input[self.read_position.get()]
    }

    fn handle_whitespace(&self) {
        while is_whitespace(self.ch.get()) {
            self.next()
        }
    }

    fn handle_single_token(&self, token_type: TokenType) -> Result<Token> {
        let token = self.tokenize(1, 0, token_type);
        self.next();

        token
    }

    fn handle_double_token(&self, token_type: TokenType) -> Result<Token> {
        let token = self.tokenize(2, 0, token_type);
        self.next();
        self.next();

        token
    }

    fn read_operator_double(&self, token_type: TokenType) -> Result<Token> {
        if self.peek() == '=' {
            match self.ch.get() {
                '=' => self.handle_double_token(TokenType::Equal),
                '!' => self.handle_double_token(TokenType::NotEqual),
                '<' => self.handle_double_token(TokenType::GreaterThanEqual),
                '>' => self.handle_double_token(TokenType::LessThanEqual),
                c @ _ => Err(TodoError(format!("unsupported token {}", c))),
            }
        } else {
            self.handle_single_token(token_type)
        }
    }

    fn read(&self) -> Result<Token> {
        self.handle_whitespace();
        match self.ch.get() {
            '=' => self.read_operator_double(TokenType::Assign),
            '+' => self.handle_single_token(TokenType::Plus),
            '-' => self.handle_single_token(TokenType::Minus),
            '!' => self.read_operator_double(TokenType::Bang),
            '*' => self.handle_single_token(TokenType::Asterisk),
            '/' => self.handle_single_token(TokenType::Slash),
            '<' => self.read_operator_double(TokenType::LessThan),
            '>' => self.read_operator_double(TokenType::GreaterThan),
            ';' => self.handle_single_token(TokenType::Semicolon),
            '(' => self.handle_single_token(TokenType::LeftParen),
            ')' => self.handle_single_token(TokenType::RightParen),
            ',' => self.handle_single_token(TokenType::Comma),
            '{' => self.handle_single_token(TokenType::LeftBrace),
            '}' => self.handle_single_token(TokenType::RightBrace),
            '[' => self.handle_single_token(TokenType::LeftBracket),
            ']' => self.handle_single_token(TokenType::RightBracket),
            ':' => self.handle_single_token(TokenType::Colon),
            '"' => self.handle_string_token(),
            '\0' => self.tokenize(0, 0, TokenType::EndOfFile),
            ch if is_identifier(ch) => self.read_identifier(),
            ch if is_digit(ch) => self.read_digit(),
            c @ _ => Err(TodoError(format!("unsupported token {}", c))),
        }
    }

    fn read_identifier(&self) -> Result<Token> {
        let position = self.position.get();
        while is_identifier(self.ch.get()) {
            self.next();
        }
        let end_position = self.position.get();
        let length = end_position - position;

        let identifier: &str = &self.input[position..end_position].iter().collect::<String>();
        match identifier {
            "let" => self.tokenize(length, length, TokenType::Let),
            "fn" => self.tokenize(length, length, TokenType::Function),
            "true" => self.tokenize(length, length, TokenType::True),
            "false" => self.tokenize(length, length, TokenType::False),
            "if" => self.tokenize(length, length, TokenType::If),
            "else" => self.tokenize(length, length, TokenType::Else),
            "return" => self.tokenize(length, length, TokenType::Return),
            _ => self.tokenize(length, length, TokenType::Identifier),
        }
    }

    fn read_digit(&self) -> Result<Token> {
        let position = self.position.get();
        while is_digit(self.ch.get()) {
            self.next()
        }
        let end_position = self.position.get();
        let length = end_position - position;

        self.tokenize(length, length, TokenType::Integer)
    }

    /// Move the processing position to the next token
    fn next(&self) {
        if self.input.len() <= self.read_position.get() {
            self.ch.set(0.into());
            self.column.set(self.column.get() + 1);
        } else {
            if self.input[self.position.get()] == '\n' {
                self.line.set(self.line.get() + 1);
                self.column.set(0)
            } else {
                self.column.set(self.column.get() + 1);
            }
            self.ch.set(self.input[self.read_position.get()])
        }

        self.position.set(self.position.get() + 1);
        self.read_position.set(self.read_position.get() + 1);
    }

    fn handle_string_token(&self) -> Result<Token> {
        let position = self.position.get();
        let start = self.column.get();

        self.next();

        let mut chars: Vec<char> = Vec::new();

        while self.ch.get() != '"' && self.ch.get() != '\0' {
            if self.ch.get() == '\\' {
                self.next();
                match self.ch.get() {
                    'n' => chars.push('\n'),
                    't' => chars.push('\t'),
                    '"' => chars.push('"'),
                    '\\' => chars.push('\\'),
                    val @ _ => return Err(InvalidCharError(self.ch.get())),
                }
            } else {
                chars.push(self.ch.get())
            }
            self.next()
        }

        let end_position = self.position.get();
        let length = end_position - position;
        let end = self.column.get();

        self.next();

        // FIXME: there is some code duplication here - some how merge this with tokenizer.
        Ok(Token {
            token_type: TokenType::String,
            literal: chars.iter().collect::<String>(),
            span: Span {
                start,
                end,
                line_start: self.line.get(),
                line_end: self.line.get(),
            },
        })
    }

    fn tokenize(&self, length: usize, offset: usize, token_type: TokenType) -> Result<Token> {
        let position = self.position.get() - offset;

        let literal = if length > 0 {
            self.input[position..position + length].to_vec()
        } else {
            vec![]
        };

        let column = self.column.get() - offset;
        let end = if length > 0 { column + length - 1 } else { column };
        Ok(Token {
            token_type,
            literal: literal.iter().collect::<String>(),
            span: Span {
                start: column,
                end,
                line_start: self.line.get(),
                line_end: self.line.get(),
            },
        })
    }
}

pub fn run_lexer(source: &str) -> Result<String> {
    let contents = fs::read_to_string(source).unwrap();
    eval_lexer(&contents)
}

pub fn eval_lexer(contents: &String) -> Result<String> {
    let lines: Vec<usize> = contents
        .chars()
        .enumerate()
        .filter(|(_, ch)| *ch == '\n')
        .map(|(idx, _)| idx)
        .collect();

    let lexer = MonkeyLexer::new(&*contents);

    let mut tokens = VecDeque::new();
    let mut stop_loop = false;
    while !stop_loop {
        let result = lexer.token();
        if let Ok(token) = result {
            stop_loop = token.token_type == TokenType::EndOfFile;
            tokens.push_back(token);
        } else if let Err(message) = result {
            return Err(message);
        }
    }

    let snapshot = token_snapshot(&contents, &lines, &tokens);

    Ok(snapshot)
}

pub fn token_snapshot(program: &String, lines: &Vec<usize>, tokens: &VecDeque<Token>) -> String {
    let mut output = String::new();
    let mut working_line = usize::MAX;
    for token in tokens {
        if token.span.line_start != token.span.line_end {
            panic!("multiline command not supported ðŸ˜ž")
        }

        if token.span.line_start != working_line {
            working_line = token.span.line_start;
            let start = if working_line != 0 {
                lines[working_line - 1]
            } else {
                0
            };

            let end = if working_line < lines.len() {
                lines[working_line]
            } else {
                program.len() - 1
            };
            output += &program[start..=end];
        }

        output += &" ".repeat(token.span.start);
        output += &"^".repeat(token.span.end + 1 - token.span.start);
        output += &format!(" {token:?}");
        output += &"\n";
    }

    output
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_lexer() {
        let result = run_lexer("monkey/lexer.mky").unwrap();

        insta::assert_snapshot!(result)
    }
}
