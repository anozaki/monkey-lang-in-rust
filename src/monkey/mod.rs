use crate::monkey::error::Error;

pub mod error;

pub mod lexer;
pub mod helper;
pub mod parser;
pub mod token;


pub mod interpreter;

pub type Result<T> = core::result::Result<T, Error>;
