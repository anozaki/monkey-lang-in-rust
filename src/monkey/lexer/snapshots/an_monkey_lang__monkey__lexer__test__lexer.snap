---
source: src/monkey/lexer/mod.rs
expression: result
---
let five = 5;
^^^ Token(Let, "let" [L0-0:2])
    ^^^^ Token(Identifier, "five" [L0-4:7])
         ^ Token(Assign, "=" [L0-9:9])
           ^ Token(Integer, "5" [L0-11:11])
            ^ Token(Semicolon, ";" [L0-12:12])

let ten = 10;
^^^ Token(Let, "let" [L1-0:2])
    ^^^ Token(Identifier, "ten" [L1-4:6])
        ^ Token(Assign, "=" [L1-8:8])
          ^^ Token(Integer, "10" [L1-10:11])
            ^ Token(Semicolon, ";" [L1-12:12])

let add = fn(x, y) {
^^^ Token(Let, "let" [L3-0:2])
    ^^^ Token(Identifier, "add" [L3-4:6])
        ^ Token(Assign, "=" [L3-8:8])
          ^^ Token(Function, "fn" [L3-10:11])
            ^ Token(LeftParen, "(" [L3-12:12])
             ^ Token(Identifier, "x" [L3-13:13])
              ^ Token(Comma, "," [L3-14:14])
                ^ Token(Identifier, "y" [L3-16:16])
                 ^ Token(RightParen, ")" [L3-17:17])
                   ^ Token(LeftBrace, "{" [L3-19:19])

  x + y;
  ^ Token(Identifier, "x" [L4-2:2])
    ^ Token(Plus, "+" [L4-4:4])
      ^ Token(Identifier, "y" [L4-6:6])
       ^ Token(Semicolon, ";" [L4-7:7])

};
^ Token(RightBrace, "}" [L5-0:0])
 ^ Token(Semicolon, ";" [L5-1:1])

let result = add(five, ten);
^^^ Token(Let, "let" [L7-0:2])
    ^^^^^^ Token(Identifier, "result" [L7-4:9])
           ^ Token(Assign, "=" [L7-11:11])
             ^^^ Token(Identifier, "add" [L7-13:15])
                ^ Token(LeftParen, "(" [L7-16:16])
                 ^^^^ Token(Identifier, "five" [L7-17:20])
                     ^ Token(Comma, "," [L7-21:21])
                       ^^^ Token(Identifier, "ten" [L7-23:25])
                          ^ Token(RightParen, ")" [L7-26:26])
                           ^ Token(Semicolon, ";" [L7-27:27])

!-/*5;
^ Token(Bang, "!" [L8-0:0])
 ^ Token(Minus, "-" [L8-1:1])
  ^ Token(Slash, "/" [L8-2:2])
   ^ Token(Asterisk, "*" [L8-3:3])
    ^ Token(Integer, "5" [L8-4:4])
     ^ Token(Semicolon, ";" [L8-5:5])

5 < 10 > 5;
^ Token(Integer, "5" [L9-0:0])
  ^ Token(LessThan, "<" [L9-2:2])
    ^^ Token(Integer, "10" [L9-4:5])
       ^ Token(GreaterThan, ">" [L9-7:7])
         ^ Token(Integer, "5" [L9-9:9])
          ^ Token(Semicolon, ";" [L9-10:10])

if (5 < 10) {
^^ Token(If, "if" [L11-0:1])
   ^ Token(LeftParen, "(" [L11-3:3])
    ^ Token(Integer, "5" [L11-4:4])
      ^ Token(LessThan, "<" [L11-6:6])
        ^^ Token(Integer, "10" [L11-8:9])
          ^ Token(RightParen, ")" [L11-10:10])
            ^ Token(LeftBrace, "{" [L11-12:12])

    return true;
    ^^^^^^ Token(Return, "return" [L12-4:9])
           ^^^^ Token(True, "true" [L12-11:14])
               ^ Token(Semicolon, ";" [L12-15:15])

} else {
^ Token(RightBrace, "}" [L13-0:0])
  ^^^^ Token(Else, "else" [L13-2:5])
       ^ Token(LeftBrace, "{" [L13-7:7])

    return false;
    ^^^^^^ Token(Return, "return" [L14-4:9])
           ^^^^^ Token(False, "false" [L14-11:15])
                ^ Token(Semicolon, ";" [L14-16:16])

}
^ Token(RightBrace, "}" [L15-0:0])

5 == 4;
^ Token(Integer, "5" [L17-0:0])
  ^^ Token(Equal, "==" [L17-2:3])
     ^ Token(Integer, "4" [L17-5:5])
      ^ Token(Semicolon, ";" [L17-6:6])

5 != 4;
^ Token(Integer, "5" [L18-0:0])
  ^^ Token(NotEqual, "!=" [L18-2:3])
     ^ Token(Integer, "4" [L18-5:5])
      ^ Token(Semicolon, ";" [L18-6:6])

5 <= 6;
^ Token(Integer, "5" [L19-0:0])
  ^^ Token(GreaterThanEqual, "<=" [L19-2:3])
     ^ Token(Integer, "6" [L19-5:5])
      ^ Token(Semicolon, ";" [L19-6:6])

5 >= 7;
^ Token(Integer, "5" [L20-0:0])
  ^^ Token(LessThanEqual, ">=" [L20-2:3])
     ^ Token(Integer, "7" [L20-5:5])
      ^ Token(Semicolon, ";" [L20-6:6])

fn() {
^^ Token(Function, "fn" [L22-0:1])
  ^ Token(LeftParen, "(" [L22-2:2])
   ^ Token(RightParen, ")" [L22-3:3])
     ^ Token(LeftBrace, "{" [L22-5:5])

}
^ Token(RightBrace, "}" [L24-0:0])

"this is a test";
^^^^^^^^^^^^^^^^ Token(String, "this is a test" [L26-0:15])
                ^ Token(Semicolon, ";" [L26-16:16])

let a = "this is another\n\"test\"";
^^^ Token(Let, "let" [L27-0:2])
    ^ Token(Identifier, "a" [L27-4:4])
      ^ Token(Assign, "=" [L27-6:6])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^ Token(String, "this is another\n\"test\"" [L27-8:34])
                                   ^ Token(Semicolon, ";" [L27-35:35])

[1, 2, 3];
^ Token(LeftBracket, "[" [L28-0:0])
 ^ Token(Integer, "1" [L28-1:1])
  ^ Token(Comma, "," [L28-2:2])
    ^ Token(Integer, "2" [L28-4:4])
     ^ Token(Comma, "," [L28-5:5])
       ^ Token(Integer, "3" [L28-7:7])
        ^ Token(RightBracket, "]" [L28-8:8])
         ^ Token(Semicolon, ";" [L28-9:9])

[1, 2 * 3, 5];
^ Token(LeftBracket, "[" [L29-0:0])
 ^ Token(Integer, "1" [L29-1:1])
  ^ Token(Comma, "," [L29-2:2])
    ^ Token(Integer, "2" [L29-4:4])
      ^ Token(Asterisk, "*" [L29-6:6])
        ^ Token(Integer, "3" [L29-8:8])
         ^ Token(Comma, "," [L29-9:9])
           ^ Token(Integer, "5" [L29-11:11])
            ^ Token(RightBracket, "]" [L29-12:12])
             ^ Token(Semicolon, ";" [L29-13:13])

{foo: 5, a: "test"}
^ Token(LeftBrace, "{" [L31-0:0])
 ^^^ Token(Identifier, "foo" [L31-1:3])
    ^ Token(Colon, ":" [L31-4:4])
      ^ Token(Integer, "5" [L31-6:6])
       ^ Token(Comma, "," [L31-7:7])
         ^ Token(Identifier, "a" [L31-9:9])
          ^ Token(Colon, ":" [L31-10:10])
            ^^^^^^ Token(String, "test" [L31-12:17])
                  ^ Token(RightBrace, "}" [L31-18:18])
                    ^ Token(EndOfFile, "" [L31-20:20])

